node/pkg/gwrelayer/gwrelayer.go

```diff
@@ -0,0 +1,274 @@
+	} else {
+		if err := supervisor.Run(ctx, "gwrworker", common.WrapWithScissors(gwr.worker, "gwrworker")); err != nil {
+			return fmt.Errorf("failed to start submit vaa worker: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Close closes the connection to the smart contract.
+func (gwr *GatewayRelayer) Close() {
+	if gwr.wormchainConn != nil {
+		gwr.wormchainConn.Close()
+		gwr.wormchainConn = nil
+	}
+}
+
+// convertBech32AddressToWormhole converts a bech32 address to a wormhole address.
+func convertBech32AddressToWormhole(contractAddress string) (vaa.Address, error) {
+	addrBytes, err := sdktypes.GetFromBech32(contractAddress, "wormhole")
+	if err != nil {
+		return vaa.Address{}, fmt.Errorf(`failed to decode target address "%s": %w`, contractAddress, err)
+	}
+	return vaa.Address(addrBytes), nil
+}
+
+// SubmitVAA checks to see if the VAA should be submitted to the smart contract, and if so, writes it to the channel for publishing.
+func (gwr *GatewayRelayer) SubmitVAA(v *vaa.VAA) {
+	if shouldPub, err := shouldPublish(v.Payload, vaa.ChainIDWormchain, gwr.targetAddress); err != nil {
+		gwr.logger.Error("failed to check if vaa should be published", zap.String("msgId", v.MessageID()), zap.Error(err))
+		return
+	} else if !shouldPub {
+		return
+	}
+
+	select {
+	case gwr.subChan <- v:
+		gwr.logger.Debug("submitted vaa to channel", zap.String("msgId", v.MessageID()))
+	default:
+		channelFullErrors.Inc()
+		gwr.logger.Error("unable to submit vaa because the channel is full, dropping it", zap.String("msgId", v.MessageID()))
+	}
+}
+
+// shouldPublish returns true if a message should be forwarded to the contract on wormchain, false if not.
+func shouldPublish(payload []byte, targetChain vaa.ChainID, targetAddress vaa.Address) (bool, error) {
+	if len(payload) == 0 {
+		return false, nil
+	}
+
+	if payload[0] != 3 {
+		return false, nil
+	}
+
+	hdr, err := vaa.DecodeTransferPayloadHdr(payload)
+	if err != nil {
+		return false, fmt.Errorf("failed to decode payload: %w", err)
+	}
+
+	if hdr.TargetChain != targetChain || hdr.TargetAddress != targetAddress {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// worker listens for VAAs and submits them to the smart contract.
+func (gwr *GatewayRelayer) worker(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case v := <-gwr.subChan:
+			if err := gwr.submitVAAToContract(v); err != nil {
+				gwr.logger.Error("failed to submit vaa to contract", zap.String("msgId", v.MessageID()), zap.Error(err))
+				return fmt.Errorf("failed to submit vaa to contract: %w", err)
+			}
+		}
+	}
+}
+
+// submitVAAToContract submits a VAA to the smart contract on wormchain.
+func (gwr *GatewayRelayer) submitVAAToContract(v *vaa.VAA) error {
+	_, err := SubmitVAAToContract(gwr.ctx, gwr.logger, gwr.wormchainConn, gwr.contractAddress, v)
+	if err != nil {
+		submitErrors.Inc()
+		return err
+	}
+	// TODO: Need to check txResp for "VAA already submitted", which should not be an error.
```

Comment:
eta for this todo?

Reply:
Once I can talk to the real contract and see how it behaves. :grin: